### PR TITLE
Add completion for :back/:forward

### PR DESCRIPTION
**Pull request by toofar**
Adds a completion for :back and :forward based on the history API for
the current tab in the current window. The column widths are copied from
the buffer completion. The first column is just the index in history.

I imagine that people may be interested in the `item.lastVisited()`
value which appears to be supported by both backends but we don't save
it out to the session file and it isn't used anywhere else in the
codebase so I held off on that.

Both backend's history classes also have `backItems()` and
`forwardItems()` but they aren't exposed in qutebrowsers history API,
hence the conditional in the list comprehension.

I haven't added a delete function, although I'm sure people will want
it, because:

* since there is no API for it it would involve serializing, editing,
  and deserealizing the history which would cause the tab to be reloaded at
  which point you may as well just have opened a new one and closed the
  tab with the unwanted history item
* it would only remove from the tab history, not the global history
  which may be surprising behavior

I had to pass `index=index` in the command methods because `forward`
isn't defined as a keyword argument in `_back_forward` but it is called
as one.

Passing negative values for index will do slightly weird things (for
back it go all the way back, for forward it'll go 1 forward) but I don't
care enough to add an extra check in there.

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/4180)

### CODE REVIEW

**Comment by toofar** [comment]
I changed qutebrowser/completion/models/miscmodels.py and it didn't request a review from rcorre but I changed tests/unit/completion/test_models.py and it did. Maybe the `qutebrowser/completion/*  @rcorre` line in `.github/CODEOWNERS` should be changed to have `**` instead so it is recursive? https://help.github.com/articles/about-codeowners/ says it uses the same format as `.gitignore` and https://git-scm.com/docs/gitignore#_pattern_format says that is a thing.

**Comment by toofar** [comment]
I'm not sure of the best way to handle those lint failures.

**Comment by rcorre** [comment]
> aybe the qutebrowser/completion/* @rcorre line in .github/CODEOWNERS should be changed to have ** instead so it is recursive?

I thought we had done this before, but I guess not. We should though.

**Comment by The-Compiler** [comment]
Huh, me too. Guess we talked about it and then forgot :wink: I did so in 5f10c9c15f534a51fd3896335b051462a56eab89 now.

File: qutebrowser/browser/commands.py [hunk 1]
```diff
@@ -552,6 +552,17 @@ def _back_forward(self, tab, bg, window, count, forward):
         else:
             widget = self._current_widget()
 
+        if not count and index:
+            try:
+                index = int(index)
+            except ValueError:
+                raise cmdexc.CommandError("Invalid int value {}"
```
Comment (rcorre) [inline qutebrowser/browser/commands.py]: How about "index must be an int" or something so the user knows which argument they messed up?

Comment (toofar) [reply qutebrowser/browser/commands.py]: Okay, although that is the only non-flag argument back and forward take.

**Comment by toofar** [comment]
I added `{back,forward}_items` to the tab history API but I don't know a good way to test them. Since we can't instantiate a QWebEngineHistory object (because it is just a facade over some chromium stuff I guess) we can't really test it a full tab and navigating to some pages or using `history.load_items()` with some `sessions.TabHistoryItems`, which also triggers a page load if one of the items is marked as active.

**Comment by The-Compiler** [comment]
It's quite odd that `QDateTime` doesn't support a strftime-compatible format... However, you can turn it into a Python [datetime](https://docs.python.org/3/library/datetime.html) object and use that:

```python
>>> from PyQt5.QtCore import QDateTime
>>> import datetime

>>> dt = QDateTime.currentDateTime()
>>> dt
PyQt5.QtCore.QDateTime(2018, 9, 28, 20, 5, 34, 381)

>>> pydt = datetime.datetime.fromtimestamp(dt.toSecsSinceEpoch())
>>> pydt
datetime.datetime(2018, 9, 28, 20, 5, 34)

>>> pydt.strftime('%Y-%m-%d')
'2018-09-28'
```

File: qutebrowser/browser/commands.py [hunk 2]
```diff
@@ -552,6 +552,12 @@ def _back_forward(self, tab, bg, window, count, forward):
         else:
             widget = self._current_widget()
 
+        if not count and index:
+            count = abs(history.current_idx() - index)
+
+        if not count:
+            count = 1
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: I found this a bit hard to follow, and it'll also mean we go back one item when `:back` is used with the current index for some reason.

What about this instead?

```python
if count is None:
    if index is None:
        count = 1
    else:
        count = abs(history.current_idx() - index)
```

File: qutebrowser/browser/webengine/webenginetab.py [hunk 3]
```diff
@@ -514,6 +514,12 @@ def can_go_back(self):
     def can_go_forward(self):
         return self._history.canGoForward()
 
+    def back_items(self, max_items: int = 99999):
+        return self._history.backItems(max_items)
```
Comment (The-Compiler) [inline qutebrowser/browser/webengine/webenginetab.py]: I'm not sure we should expose `max_items` at all - what about just passing `self._history.count()` there?

File: qutebrowser/misc/sessions.py [hunk 4]
```diff
@@ -387,9 +391,14 @@ def _load_tab(self, new_tab, data):
...
                 orig_url = url
+            if histentry.get("last_visited"):
+                last_visited = QDateTime.fromString(histentry.get("last_visited"))
+            else:
+                last_visited = None
             entry = TabHistoryItem(url=url, original_url=orig_url,
                                    title=histentry['title'], active=active,
-                                   user_data=user_data)
+                                   user_data=user_data,
+                                   last_visited=last_visited)
```
Comment (The-Compiler) [inline qutebrowser/misc/sessions.py]: Does this actually work as-is? Wouldn't we need to find out how that date is (de)serialized in `tabhistory.py` ([webkit](https://github.com/qutebrowser/qutebrowser/blob/master/qutebrowser/browser/webkit/tabhistory.py), [webengine](https://github.com/qutebrowser/qutebrowser/blob/master/qutebrowser/browser/webengine/tabhistory.py)) for this to work?

File: qutebrowser/misc/sessions.py [hunk 5]
```diff
@@ -179,6 +181,8 @@ def _save_tab_item(self, tab, idx, item):
             # QtWebEngine
             user_data = None
 
+        data['last_visited'] = item.lastVisited().toString()
```
Comment (The-Compiler) [inline qutebrowser/misc/sessions.py]: The default `Qt::TextDate` format is locale-dependent, so `Qt::ISODate` would probably be a better fit here.

File: qutebrowser/misc/sessions.py [hunk 6]
```diff
@@ -387,9 +391,14 @@ def _load_tab(self, new_tab, data):
                     histentry['original-url'].encode('ascii'))
             else:
                 orig_url = url
+            if histentry.get("last_visited"):
+                last_visited = QDateTime.fromString(histentry.get("last_visited"))
```
Comment (The-Compiler) [inline qutebrowser/misc/sessions.py]: ditto

File: tests/unit/completion/test_models.py [hunk 7]
```diff
@@ -23,9 +23,11 @@
 import random
 import string
 from datetime import datetime
+from unittest import mock
 
 import pytest
 from PyQt5.QtCore import QUrl
+from PyQt5.QtWebEngineWidgets import QWebEngineHistory, QWebEngineHistoryItem
```
Comment (The-Compiler) [inline tests/unit/completion/test_models.py]: This [breaks the tests](https://travis-ci.org/qutebrowser/qutebrowser/jobs/427092466) with QtWebKit (without QtWebEngine installed). I guess you could catch the `ImportError` and ignore it here, and in the tests do `pytest.importorskip('PyQt5.QtWebEngineWidgets')`.

File: tests/unit/completion/test_models.py [hunk 8]
```diff
@@ -939,3 +941,73 @@ def bench():
...
+
+    history = []
+    for url, title in [
+            ("http://example.com/index", "list of things"),
+            ("http://example.com/thing1", "thing1 detail"),
+            ("http://example.com/thing2", "thing2 detail"),
+            ("http://example.com/thing3", "thing3 detail"),
+            ("http://example.com/thing4", "thing4 detail"),
+    ]:
+        entry = mock.Mock(spec=QWebEngineHistoryItem)
```
Comment (The-Compiler) [inline tests/unit/completion/test_models.py]: Hmm, Qt [doesn't seem to like that](https://travis-ci.org/qutebrowser/qutebrowser/jobs/427092467).

File: qutebrowser/misc/sessions.py (hunk 4 above)
Comment (toofar) [reply qutebrowser/misc/sessions.py]: It was not working. Chromium wants the timestamp as the number of microseconds since 1601-01-01. So [this](https://github.com/qutebrowser/qutebrowser/blob/6f028e9/qutebrowser/browser/webengine/tabhistory.py#L63) would have always been in January 1601.

Comment (toofar) [reply qutebrowser/misc/sessions.py]: And `QWebHistoryItem::lastVisited()` appears to always return `QDateTime()`... It's currently falling back to 1970-01-01 for invalid (and earlier than 1970) dates so I think I will just leave that as is for webkit. [Here](https://bugs.webkit.org/show_bug.cgi?id=60940) is an old bug (marked resolved invalid) that says something similar (presumably with a local timezone adjustment).

**Comment by arza-zara** [comment]
I have been using this for two months without problems.

As a future improvement it would be nice to complete both backward and forward history at the same time and maybe add the completion to `:open` like #4955.

File: qutebrowser/completion/models/miscmodels.py [hunk 9]
```diff
@@ -167,3 +168,57 @@ def window(*, info):
     model.add_category(listcategory.ListCategory("Windows", windows))
 
     return model
+
+def _qdatetime_to_completion_format(qdate):
+    if not qdate.isValid():
+        ts = 0
+    else:
+        ts = qdate.toSecsSinceEpoch()
```
[the remark this answers is not in the document]
Comment (The-Compiler) [reply qutebrowser/completion/models/miscmodels.py]: I prefer the current way personally - it's clearer what the intention behind it is.

File: qutebrowser/browser/webengine/tabhistory.py [hunk 10]
```diff
@@ -60,7 +58,14 @@ def _serialize_item(item, stream):
...
     stream.writeBool(False)
     ## static_cast<qint64>(entry->GetTimestamp().ToInternalValue());
-    stream.writeInt64(int(time.time()))
+    if item.last_visited is None:
+        unix_msecs = 0
+    else:
+        unix_msecs = item.last_visited.toMSecsSinceEpoch()
+    # 11644516800000 is the number of milliseconds from
+    # 1601-01-01T00:00 (Windows NT Epoch) to 1970-01-01T00:00 (UNIX Epoch)
+    nt_usecs = (unix_msecs + 11644516800000) * 1000
```
[the remark this answers is not in the document]
Comment (The-Compiler) [reply qutebrowser/browser/webengine/tabhistory.py]: There is a comment though? :wink:

[the remark this answers is not in the document]
Comment (rcorre) [reply qutebrowser/browser/webengine/tabhistory.py]: There is :)

The comment just explains what the conversion is though, not why we're doing it.
Even if you understand that we're converting from unix epoch to nt, it still seems like a pretty weird thing to do.
I guess you can always check `git blame` though.

File: tests/unit/completion/test_models.py (hunk 8 above)
Comment (The-Compiler) [reply tests/unit/completion/test_models.py:1176]: Apparently this works nowadays (even with Qt 5.7), no idea why, but hey, I'm happy that way... :slightly_smiling_face:

**Comment by The-Compiler** [comment]
Merged! :tada: I resolved the conflicts, added a couple of fixes, cleanups and also completion for `:undo` while I was at it:

- f86cd440d Merge remote-tracking branch 'origin/pr/4180' into completion
- f8e83dcaa Fix merge
- 9bdcea42d Fix QDateTime conversions
- 506dadb1b Fix lint
- 3cb12fd36 Add time to completion.timestamp_format and fix docs
- 5e2275502 Restructure _update_completion a bit
- e9cf19503 Move current tab to CompletionInfo
- dc1985b84 Update changelog
- 63b44c17a Use utils.unused() everywhere
- f42cc9959 Add completion for :undo
- 8afe27885 Add timestamp to :undo completion
- 5efb66543 Update changelog

Thanks @toofar for the PR and @rcorre @arza-zara for the reviews!
